typing/printtyp.ml

```diff
@@ -1638,19 +1638,37 @@ let rec tree_of_modtype ?(ellipsis=false) = function
       Omty_alias (tree_of_path Module p)
 
 and tree_of_signature sg =
-  wrap_env (fun env -> env) (tree_of_signature_rec !printing_env false) sg
-
-and tree_of_signature_rec env' in_type_group = function
-    [] -> stop_type_group env'; []
-  | item :: rem as items ->
-      let in_type_group = still_in_type_group env' in_type_group item in
-      let (sg, rem) = filter_rem_sig item rem in
-      hide_rec_items items;
-      protect_rec_items items;
-      reset_naming_context ();
-      let trees = trees_of_sigitem item in
-      let env' = Env.add_signature (item :: sg) env' in
-      trees @ tree_of_signature_rec env' in_type_group rem
+  wrap_env (fun env -> env)
+    (tree_of_signature_rec (fun _ (_,x) -> x) !printing_env) sg
+
+and tree_of_signature_rec : 'r. (_->_->'r) -> _ -> _ -> 'r list
+  = fun decorate env' sg ->
```

Comment:
I spent a bit of time stuck trying to understand what `decorate` means besides "the thing that lets me factorize the other piece of code at the end". I believe that a better way to do this would be to change the type of `tree_of_signature_rec` (and `..._sigitem_group`) to return, instead of an `out_sig_item list`, an `(env * sig_item * out_sig_item) list`, and let the two callers map on this result for post-processing.

Reply:
I used a slightly more structured output: an  ` (env * (sig_item * out_sig_item) list) list)`.